### PR TITLE
Add struct and functionality for using user input 

### PR DESCRIPTION
**Pull request by rbehjati**
Ref #1191

Builds on top of #1386 and adds struct and functionality for parsing and using user-provided input. 

Moves `InputOptions` to package `pkg` to be able to pass it as an argument to `NewDockerBuildConfig` without introducing a circular dependency.

### CODE REVIEW

**Review by laurentsimon — APPROVED** [review]

**Review by ianlewis — APPROVED** [review]

File: internal/builders/docker/pkg/config_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+				Alg: "sha256",
+				Value: "9e2ba52487d945504d250de186cb4fe2e3ba023ed2921dd6ac8b97ed43e76af9",
+			},
+		},
+		BuildConfigPath: buildConfigPath,
+	}
+
+	if !cmp.Equal(*got, want) {
+		t.Errorf(cmp.Diff(*got, want))
+	}
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply internal/builders/docker/pkg/config_test.go]: Thanks. Fixed.

File: internal/builders/docker/testdata/config.toml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Simple command for generating a file.
+command = ["cp", "internal/builders/docker/testdata/config.toml", "config.toml"]
+# Path to the file generated by the command above.
+artifact_path = "config.toml"
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply internal/builders/docker/testdata/config.toml]: Done.

File: internal/builders/docker/pkg/config.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+)
+
+// BuildConfig is a collection of parameters to use for building the artifact.
+type BuildConfig struct {
+	// TODO(#1191): Add env and options if needed. 
+	// Command to pass to `docker run`. The command is taken as an array
+	// instead of a single string to avoid unnecessary parsing. See
+	// https://docs.docker.com/engine/reference/builder/#cmd and
+	// https://man7.org/linux/man-pages/man3/exec.3.html for more details.
+	Command []string `toml:"command"`
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply internal/builders/docker/pkg/config.go:44]: Done.

File: internal/builders/docker/pkg/config.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+// ToMap returns this instance as a mapping between the algorithm and value.
+func (d* Digest) ToMap() map[string]string {
+	return map[string]string{d.Alg: d.Value}
+}
+
+// LoadBuildConfigFromFile loads build configuration from a toml file in the given path and returns an instance of BuildConfig.
+func LoadBuildConfigFromFile(path string) (*BuildConfig, error) {
+	tomlTree, err := toml.LoadFile(path)
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply internal/builders/docker/pkg/config.go:166]: Thanks. Replaced the validation check above that only checked that the path is relative with a call to `utils.PathIsUnderCurrentDirectory`.

**Review by rbehjati — COMMENTED** [review]
Thanks for the review.

File: internal/builders/docker/commands.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+			check(err)
+			log.Printf("The config is: %v\n", config)
+
+			// TODO(#1191): Set up build state using config, and build the artifact.
+			artifacts := "To be implemented"
+			log.Printf("Generated artifacts are: %v\n", artifacts)
+		},
+	}
+
+    cmd.Flags().StringVarP(&buildConfigPath, "build_config_path", "c", "",
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply internal/builders/docker/commands.go]: Done.

File: internal/builders/docker/pkg/config.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+type DockerBuildConfig struct {
+	SourceRepo      string
+	SourceDigest    Digest
+	BuilderImage    DockerImage
+	BuildConfigPath string
+}
+
+// NewDockerBuildConfig validates the inputs and generates an instance of
+// DockerBuildConfig.
+func NewDockerBuildConfig(io *InputOptions) (*DockerBuildConfig, error) {
```
Comment (rbehjati) [inline internal/builders/docker/pkg/config.go:72]: I could make the fields in `DockerBuildConfig` private/unexposed to make sure it is not possible to create an instance of it without performing these checks. What do you think? @asraa & @ianlewis (somewhat related to the comment below on validating the path)

**Comment by rbehjati** [comment]
@asraa @laurentsimon @ianlewis I am happy with this PR. Can you please merge it?

**Comment by rbehjati** [comment]
> Can you rebase and solve the conflicts?

Done. But now the CI step for `verify base image` fails. I don't think that is due to my changes. Is it?

**Comment by laurentsimon** [comment]
No it's not your code. @ianlewis thoughts on what's going on?
I'm not sure there is a commitment from the image maintainers to sign the images. I'll merge this PR to unblock you.
Filed https://github.com/slsa-framework/slsa-github-generator/issues/1401 to resolve later.

**Comment by rbehjati** [comment]
Many thanks :)

**Pull request merged**